Fix wrapper subclass reentrant dispatch + TorchDispatchMode

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134683
* __->__ #136566

Fixes #136565

This PR makes the python fallback robust to the case where there are no active modes & no tensors with the Python key. In this case, simply redispatch with the Python key disabled.

This was found when trying to use reentrant dispatch for NJT to get decompositions under `inference_mode()` when the autograd key is disabled.

jbschlosser committed:
Fix wrapper subclass reentrant dispatch + TorchDispatchMode

[ghstack-poisoned]

soulitzer commented on aten/src/ATen/core/PythonFallbackKernel.cpp:
```diff
@@ -82,14 +88,33 @@ void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+      (*interpreter)->dispatch(op, stack);
+    } else {
+      // At this point, there are no modes in the stack and no tensors with the python key.
+      // so remove and disable the python key before redispatching.
```
Could be useful to add a link to the issue here just to point out when this might happen

soulitzer commented on aten/src/ATen/core/PythonFallbackKernel.cpp:
```diff
@@ -82,14 +88,33 @@ void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+      // so remove and disable the python key before redispatching.
+      c10::impl::ExcludeDispatchKeyGuard no_python_guard(c10::DispatchKey::Python);
+      c10::DispatchKeySet keyset = dispatch_keys.remove(c10::DispatchKey::Python);
+      op.redispatchBoxed(keyset, stack);
```
Is this equivalent to redispatch using `after_Python_keyset`?

jbschlosser replied:
I may be wrong but I don't think so - IIUC `after_Python_keyset` is literally everything possible after Python, whereas we want to redispatch using only the *enabled* ones after Python.

bdhirsh replied:
yep that's right (this says "take the exacted DispatchKeySet that the dispatcher computed before invoking the Python key, then remove the Python key and redispatch)

soulitzer commented on aten/src/ATen/core/PythonFallbackKernel.cpp:
```diff
@@ -82,14 +88,33 @@ void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+    } else {
+      // At this point, there are no modes in the stack and no tensors with the python key.
+      // so remove and disable the python key before redispatching.
+      c10::impl::ExcludeDispatchKeyGuard no_python_guard(c10::DispatchKey::Python);
```
I guess alternatively, instead of adding it from the exclude set, we could remove it from the include set, and that would make it possible for someone in theory to push another mode somewhere else (though that seems unlikely?)

bdhirsh replied:
I think the exclude guard here is actually... unnecessary? (does the test fail if you remove it?)

(1) `redispatchBoxed() will ignore the current include/exclude TLS, and just directly dispatch to the next key in the keyset that you passed in (which is almost guaranteed here to be one of the backend keys, CPU/CUDA/BackendSelect)

(2) The TLS will only matter in the case where the next function the dispatcher hits is composite, and invokes other ATen ops. Although... if we hit another op and end up going back to the python key, we should end up just re-triggering this logic.

jbschlosser replied:
If I remove the exclude guard, I run into:
```
RuntimeError: 0 INTERNAL ASSERT FAILED at "/home/jbschlosser/branches/misc_fixes/aten/src/ATen/core/PythonFallbackKernel.cpp":123, please report a bug to PyTorch. Hit Python dispatch key but no arguments had PyInterpreter (no tensor args?)
```

and I don't quite understand why. are you sure the TLS is ignored by `redispatchBoxed()`?

jbschlosser replied:
okay, clone() is composite explicit, so without the exclude guard, we'll redispatch into the python fallback for `empty_strided()` (which has no tensor inputs and thus no interpreters).

jbschlosser replied:
Updated to remove from the include set instead!

bdhirsh commented on aten/src/ATen/core/PythonFallbackKernel.cpp:
```diff
@@ -82,14 +88,33 @@ void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+        }
+
+        auto* t_interpreter = t->pyobj_slot()->pyobj_interpreter();
+        if (!interpreter && t_interpreter) {
```
nit: right now you are calling `t->pyobj_slot()->pyobj_interpreter()` on every single tensor arg, but you can fastpath and only grab it in the case when `interpreter` isn't already set.

(I know some folks are worried about torch_dispatch related cpu overhead - so I'm not sure how much this matters but anything we can do to avoid unnecessary work in this codepath seems probably worth doing

jbschlosser replied:
agree, I didn't originally realize it was expensive to query for the interpreter, but it makes sense. moved :)

jbschlosser committed:
Update on "Fix wrapper subclass reentrant dispatch + TorchDispatchMode"


Fixes #136565

This PR makes the python fallback robust to the case where there are no active modes & no tensors with the Python key. In this case, simply redispatch with the Python key disabled.

This was found when trying to use reentrant dispatch for NJT to get decompositions under `inference_mode()` when the autograd key is disabled.

[ghstack-poisoned]

bdhirsh reviewed (approved):
nice find :)

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge -i
